### ISSUE TITLE
Refactor Reseedable

### ISSUE THREAD

**Issue by serin-delaunay** [Wangscape/Wangscape#57]
Currently `Reseedable` is a struct containing a `shared_ptr` to a `Module` and a `std::function` which (in theory) either does nothing or calls the `Module`'s `SetSeed` method.

The default factories for `Reseedable`s assign the correct `SetSeed` methods, but they do it by manual template specialisation of each individual `Module` subclass with a `SetSeed` method. Also, it would be trivial to invalidate the object by changing `setSeed` to a different function.

`Reseedable` should be replaced with a set of classes that:
* Has a template parameter that can be instantiated with any `Module` subclass.
* Contains an instance of or inherits from the template parameter.
* Has a common base class which:
  * Has a virtual `setSeed` method which ensures the correct behaviour.
  * Provides base reference access to the associated `Module` instance.

TODO moved into #73.

**Comment by serin-delaunay** [comment]
I've tried to do this using derived classes, according to the following scheme:
```
            Module
           /      \
ReseedableBase     M = Perlin, Billow, ...
           \      /
         Reseedable<M>
```
Code:
```
class ReseedableBase : public virtual Module
{
public:
    virtual void SetSeed(int seed) = 0;
};

template<typename M>
class ReseedableT :
    public virtual ReseedableBase,
    public virtual M
{
public:
    ReseedableT() :
        M::Module(GetSourceModuleCount()),
        M()
    { }
    virtual int GetSourceModuleCount() const
    {
        return M::GetSourceModuleCount();
    }
    virtual double GetValue(double x, double y, double z) const
    {
        return M::GetValue(x, y, z);
    }
    virtual void SetSeed(int seed);
};

template<typename M> void ReseedableT<M>::SetSeed(int seed)
{ }

template<> void ReseedableT<ModuleGroup>::SetSeed(int seed)
{
    ModuleGroup::SetSeed(seed);
}

// Other modules which have a SetSeed function...
```
This compiles, but if I understand virtual base classes correctly, the resulting `Reseedable<M>` classes will all have two copies of the `Module` base class. This happens despite my `ReseedableBase` and `ReseedableT<M>` classes using virtual inheritance, because all the modules in libnoise inherit non-virtually from `Module`. This can be fixed in Wangscape's module classes, but not in libnoise without forking. So now there are two choices:

1. Use a fork of libnoise where all modules inherit virtually from `Module`, and use this code. This isn't too far-fetched; we are already planning to make libnoise a submodule.
2. Make the module a member of `Reseedable`, but template `Reseedable` on the module type and replace the `SetSeed` member object with a member function whose implementation depends on the template parameter.

In any case, care must be taken to make sure that base-type `Reseedables` correctly call the derived class's `SetSeed` function when possible, and to allow constructors with arguments to be used.

**Comment by serin-delaunay** [comment]
Closed in #83.
